Add synonyms, 2ndary IDs, XREFs to Reagent

chris-grove opened the pull request:
The Construct class needs synonyms, secondary IDs and cross references. These have now been added to the Reagent and ReagentDTO superclasses to accommodate the Construct and Antibody classes (FlyBase currently submits these properties for Construct submissions to the FMS).

chris-grove commented:
FWIW, I'm throwing in a sample set of entries from the current FB CONSTRUCT submission to the FMS:
`{
  "data": [
    {
      "crossReferences": [
        {
          "id": "FB:FBtp0000078",
          "pages": [
            "construct"
          ]
        }
      ],
      "name": "P{hsneo}",
      "primaryId": "FB:FBtp0000078",
      "synonyms": [
        "P{ori ampR neoR[Hsp70Ba.PS]=hsneo}",
        "P{ori ampR neoR<sup>Hsp70Ba.PS</sup>=hsneo}",
        "P{hsneo}",
        "P{ori ampR neoR[Hsp70Bb.PS]=hsneo}",
        "P{ori ampR neoR<sup>Hsp70Bb.PS</sup>=hsneo}",
        "pUChsneo transposon"
      ]
    },
    {
      "constructComponents": [
        {
          "componentID": "FB:FBgn0001077",
          "componentRelation": "expresses",
          "componentSymbol": "ftz"
        },
        {
          "componentID": "FB:FBgn0286924",
          "componentRelation": "is_regulated_by",
          "componentSymbol": "Hsp70"
        }
      ],
      "crossReferences": [
        {
          "id": "FB:FBtp0000233",
          "pages": [
            "construct"
          ]
        }
      ],
      "name": "P{hsp70-ftz}",
      "primaryId": "FB:FBtp0000233",
      "synonyms": [
        "P{ftz[hs.PS] Adh[+t3.2]=hsp70-ftz}",
        "P{ftz<sup>hs.PS</sup> Adh<sup>+t3.2</sup>=hsp70-ftz}",
        "HS-Dm-ftz",
        "hsp70-ftz",
        "hsf2"
      ]
    },
    {
      "constructComponents": [
        {
          "componentID": "RGD:3044",
          "componentRelation": "expresses",
          "componentSymbol": "Map2"
        },
        {
          "componentRelation": "is_regulated_by",
          "componentSymbol": "UAS"
        }
      ],
      "crossReferences": [
        {
          "id": "FB:FBtp0001429",
          "pages": [
            "construct"
          ]
        }
      ],
      "name": "P{UAS-MAP2.A}",
      "primaryId": "FB:FBtp0001429",
      "secondaryIds": [
        "FB:FBtp0000433"
      ],
      "synonyms": [
        "P{UAS-Map2c.A}",
        "UAS-MAP2c",
        "UAS-M2C48C (MAP2c)",
        "P{UAS-MAP2c.A}",
        "P{w<sup>+mC</sup> Map2<sup>UAS.cAa</sup>=UAS-Map2c.A}",
        "P{w[+mC] Map2[UAS.cAa]=UAS-Map2c.A}"
      ]
    },
    {
      "constructComponents": [
        {
          "componentID": "HGNC:24650",
          "componentRelation": "expresses",
          "componentSymbol": "EHMT1"
        },
        {
          "componentRelation": "is_regulated_by",
          "componentSymbol": "UASt"
        }
      ],
      "crossReferences": [
        {
          "id": "FB:FBtp0158396",
          "pages": [
            "construct"
          ]
        }
      ],
      "name": "PBac{UAS-hEHMT1.B}",
      "primaryId": "FB:FBtp0158396",
      "synonyms": []
    },
    {
      "constructComponents": [
        {
          "componentID": "FB:FBgn0000097",
          "componentRelation": "targets",
          "componentSymbol": "aop"
        },
        {
          "componentRelation": "is_regulated_by",
          "componentSymbol": "lexAop"
        }
      ],
      "crossReferences": [
        {
          "id": "FB:FBtp0158397",
          "pages": [
            "construct"
          ]
        }
      ],
      "name": "P{lexAop-HMS01627}",
      "primaryId": "FB:FBtp0158397",
      "synonyms": [
        "Aop-atlRNAi"
      ]
    }
  ],
  "metaData": {
    "dataProvider": {
      "crossReference": {
        "id": "FB",
        "pages": [
          "homepage"
        ]
      },
      "type": "curated"
    },
    "dateProduced": "2023-03-24T09:10:32-04:00",
    "release": "2023_01"
  }
}`

chris-grove commented:
@markquintontulloch @gildossantos 
One thing to consider: If the cross-references submitted for constructs (by FlyBase or anyone else) only ever refer to IDs by the data provider, these cross-references could just go in the data provider cross references and not the class-level cross-references. Another issue is whether we want these cross references visible and/or editable from within the curation interface (a paradigm not yet currently supported for data providers in the curation system).

gildossantos reviewed (approved):
Looks good. Note that the `synonyms` slot used here accepts only a list of strings and no evidence (unlike the `allele_synonyms` or `gene_synonyms` slots, which accept a list of `NameSlotAnnotation`-type objects). There would be some data loss for FB constructs synonyms. That's fine for now, and maybe for the long term too. FB will need to figure out how imporant evidence is for our construct synonyms.

chris-grove commented:
@gildossantos 
OK, thanks! Yes, I was going to ask if you (FB) needed evidence attached to synonyms, secondary IDs, and/or cross references. We will definitely retain that data. I'm working on an update commit to address that now.

chris-grove commented:
@gildossantos 
Is it just construct synonyms you need evidence for, or also secondary IDs or cross-references?

gildossantos commented:
Sorry to keep asking @chris-grove , but in place of `name`, could we also get `construct_symbol` with range `ConstructSymbolSlotAnnotation` so that we can capture evidence for the Construct's current symbol?

chris-grove commented:
@gildossantos 
1. I've added the ConstructSynonymSlotAnnotation and other supporting parts of the model
2. The modeling for the Construct class has so far employed a 'name' slot, not a 'symbol' slot. I think we would have to have more discussions within the working group about changing the 'name' to 'symbol' and whether everyone would be OK with that. From spot checking, it seems that people might be OK with that, but we need to discuss. Whichever we go with, we'll have to create a NameSlotAnnotation property for it.

markquintontulloch commented:
> Another issue is whether we want these cross references visible and/or editable from within the curation interface (a paradigm not yet currently supported for data providers in the curation system).

@chris-grove Would we want this?  My understanding of the purpose of the dataProvider xref is to link back to the MOD page for that entity/annotation.  For new entries created in the curation system there would be no corresponding page at the MOD site, and I don't really see much of a use-case for editing the submitted entries from bulk uploads.

chris-grove commented:
@gildossantos @markquintontulloch 
I've made the discussed changes in this PR, including:
- Added 'alliance_member_release_version' slot to Ingest class
- Added construct symbol (required) and name (not required) NameSlotAnnotation properties
- Updated construct test JSON file

markquintontulloch commented:
@chris-grove Did we decide to keep crossReferences? I thought DataProvider was sufficient (I may have missed that bit of the conversation though). Also, I thought taxon was going from the Reagent level.

chris-grove commented:
> @chris-grove Did we decide to keep crossReferences? I thought DataProvider was sufficient (I may have missed that bit of the conversation though). Also, I thought taxon was going from the Reagent level.

@markquintontulloch No, we can get rid of cross references for construct. It seems they will be needed for Antibody (that was the original model specification) so I thought we could leave them in the Reagent parent class, but I'm fine if we want to push the property back down to the Antibody sublass. There are cross_reference_dtos entries in the construct JSON test file just as a theoretical, like we have for generated_by and manufactured_by (also not going to be used for Constructs). If we'd prefer to remove entries in the test file that aren't intended to be submitted, then I can certainly clean those up in the test file.

chris-grove commented:
@markquintontulloch 
Also, I want to keep taxon in the Reagent class because I think that is intended to be used by Antibody and likely other subclasses. It is already not required.

markquintontulloch commented:
Personally, I'd rather have duplicate slot definitions in child classes rather than slots in parent classes that aren't intended to be used in some child classes.  Having slots that aren't intended to be used is confusing for DQMs and developers.

markquintontulloch commented:
@chris-grove Looks good.  If we get this merged in and create a release we could get the Java updated in this sprint and avoid having to wait for another release until the updated model is ready for submission on production.

chris-grove commented:
@markquintontulloch OK, merged; good to go
